DDD-8.md

```diff
@@ -126,3 +126,146 @@ So if the snapshot is reading a chunk from a table that is modified through a su
+(3) Timestamp hw := getLastTimestamp()
+  inwindow := false
+  // other steps of event processing loop
+  while true do
+       e := next event from changelog
+       append e to outputbuffer
+       if not inwindow then
+           if e.clusterTime >= lw  //reached the low watermark
+               inwindow := true
+       
+       if e.clusterTime < hw //haven't reached the high watermark yet
+           if chunk contains e.key then
+               remove e.key from chunk
+       else //reached the high watermark
+           for each row in chunk do
+               append row to outputbuffer
+   // other steps of event processing loop
+```
+
+where _getLastTimestamp()_ is a function to get the last timestamp available in the _oplog_.
+We have found that in the _mongosh_ there is exactly this type of function [getReplicationInfo().tLast](https://www.mongodb.com/docs/manual/reference/method/db.getReplicationInfo/#mongodb-data-db.getReplicationInfo--.tLast).
+Unfortunately this is not available through the mongo driver but it easy to replicate looking at the [source code](https://github.com/mongodb/mongo/blob/274f2fddd6a0f975a678e8a953e704f7be6dda14/src/mongo/shell/db.js#L1124).
+
+It just accesses the _local_ db, every node in the RS has its own, and then query the _oplog.rs_ collection to get the max available timestamp. 
+
+First of all, this poses us a problem: the user must have permission to access the _local_ db. This, based on the different [capture.scope](https://debezium.io/documentation/reference/2.7/connectors/mongodb.html#mongodb-property-capture-scope) we support in Debezium, can be a limitation. 
+Furthermore, based on the different type of MongoDB topologies there can be problems to have a consistent information reading directly the _oplog_. 
+For example, in a scenario where the Debezium user want to connect directly to a replica, we can just read the _oplog_ of that specific node that can have inconsistent information due to the asynchronous nature of the data replication between primary and secondary nodes. 
+
+This in a RS set deployment, and it will be completely not possible in a sharded cluster deployment, since the connection goes through the _mongos_ component.
+
+To solve this limitation we can think to directly use the change stream API instead going directly on the _oplog_. 
+So we will have a change stream for the normal event streaming and a temporary change stream opened specifically to get the low and high watermarks.
+
+```java
+MongoCursor<ChangeStreamDocument<Document>> cursor = db.watch().iterator();
+ChangeStreamDocument<Document> lastDocument = cursor.next();
+
+BsonTimestamp tLast = lastDocument.getClusterTime();
+```
+
+In case no document is present we can execute a fake operation to get the operation timestamp.
+
+```java
+var database = client.getDatabase(dbName);
+var result = database.runCommand(new Document("hello", 1), BsonDocument.class);
+BsonTimestamp tLast = result.getTimestamp("operationTime");
```

Comment:
This is potentially problematic as there might be operations in the meantime. I'd suggest a different approach. 

Once change stream  cursor is iterated on -- that is when `next()` or `tryNext` was called, the cursor is guaranteed to return a resume token when `cursor.getResumeToken()` is called (see `MongoChangeStreamCursor` interface). Thus when `tryNext()` returns a document we get the resume token by calling `doc.getResumeToken` otherwise we call `cursor.getResumeToken()` (this is exactly what the our streaming does currently.